src/Microsoft.OData.Core/Json/JsonValueUtils.cs

```diff
@@ -327,11 +328,43 @@ internal static void WriteValue(TextWriter writer, byte[] value)
             else
             {
                 writer.Write(JsonConstants.QuoteCharacter);
-                writer.Write(Convert.ToBase64String(value));
+                WriteBinaryString(writer, value, ref buffer);
                 writer.Write(JsonConstants.QuoteCharacter);
             }
         }
 
+        /// <summary>
+        /// Write a byte array.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">Byte array to be written.</param>
+        /// <param name="buffer">Char buffer to use for streaming data.</param>
+        internal static void WriteBinaryString(TextWriter writer, byte[] value, ref char[] buffer)
+        {
+            Debug.Assert(writer != null, "writer != null");
+            Debug.Assert(value != null, "The value must not be null.");
+
+            buffer = BufferUtils.InitializeBufferIfRequired(buffer);
+            Debug.Assert(buffer != null);
+
+            int bufferLength = buffer.Length;
+
+            // Try to hold base64 string as much as possible in one converting.
+            int bufferByteSize = bufferLength * 3 / 4;
+
+            for (int i = 0; i < value.Length; i += bufferByteSize)

```

Comment:
>i [](start = 21, length = 1)

Consider using a more meaningful name, like "startPos"

Reply:
Thanks, I change to use "offsetIn", which is used in 'ToBase64CharArray`

Reply:
perfect